routing/matcher.go

```diff
@@ -1,111 +1,335 @@
+	param := freeWildcardRx.FindString(path)
+	if param == "" {
+		return ""
+	}
+
+	// clip '/*' and return only the name
+	return param[2:]
+}
+
+// Constructs a Matcher based on the provided definitions. If `ignoreTrailingSlash`
+// is true, the Matcher handles paths with or without a trailing slash equally.
+func newMatcher(rs []*Route, ignoreTrailingSlash bool) (*matcher, []*definitionError) {
+	var (
+		errors     []*definitionError
+		rootLeaves leafMatchers
+	)
+
+	pathMatchers := make(map[string]*pathMatcher)
+
+	for i, r := range rs {
+		l, err := newLeaf(r)
+		if err != nil {
+			errors = append(errors, &definitionError{r.Id, i, err})
+			continue
+		}
+
+		p := r.Path
+		if p == "" {
+			rootLeaves = append(rootLeaves, l)
+			continue
+		}
+
+		// normalize path
+		// in case ignoring trailing slashes, store and match all paths
+		// without the trailing slash
+		p = httppath.Clean(p)
+		if ignoreTrailingSlash && p[len(p)-1] == '/' {
+			p = p[:len(p)-1]
 		}
+
+		pm := pathMatchers[p]
+		if pm == nil {
+			pm = &pathMatcher{freeWildcardParam: freeWildcardParam(p)}
+			pathMatchers[p] = pm
+		}
+
+		pm.leaves = append(pm.leaves, l)
 	}
 
-	return matcherDefs
+	pathTree := &pathmux.Tree{}
+	for p, m := range pathMatchers {
+
+		// sort leaves during construction time, based on their priority
+		sort.Sort(m.leaves)
+
+		err := pathTree.Add(p, m)
+		if err != nil {
+			errors = append(errors, &definitionError{"", -1, err})
+		}
+	}
+
+	// sort root leaves during construction time, based on their priority
+	sort.Sort(rootLeaves)
+
+	return &matcher{pathTree, rootLeaves, ignoreTrailingSlash}, errors
 }
 
-func (r *Routing) createMatcher(defs []requestmatch.Definition) *requestmatch.Matcher {
-	m, errs := requestmatch.Make(defs, r.ignoreTrailingSlash)
-	for _, err := range errs {
-		// individual matcher entry errors are accepted here
-		log.Println(err)
+func matchPathTree(tree *pathmux.Tree, path string) (leafMatchers, map[string]string) {
+	v, params := tree.Lookup(path)
+	if v == nil {
+		return nil, nil
 	}
 
-	return m
+	// prepend slash in case of free form wildcards path segments (`/*name`),
+	pm := v.(*pathMatcher)
+	if pm.freeWildcardParam != "" {
+		freeParam := params[pm.freeWildcardParam]
+		freeParam = "/" + freeParam
+		params[pm.freeWildcardParam] = freeParam
+	}
+
+	return pm.leaves, params
 }
 
-func (r *Routing) processData(data string) (*requestmatch.Matcher, error) {
-	eskipDefs, err := eskip.Parse(data)
-	if err != nil {
-		return nil, err
+func matchRegexps(rxs []*regexp.Regexp, s string) bool {
+	for _, rx := range rxs {
+		if !rx.MatchString(s) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func matchHeader(h http.Header, key string, check func(string) bool) bool {
+	vals, has := h[key]
+	if !has {
+		return false
+	}
+
+	for _, val := range vals {
+		if check(val) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchHeaders(exact map[string]string, hrxs map[string][]*regexp.Regexp, h http.Header) bool {
```

Comment:
good question. there can be some performance benefits, though, but doubt that any significant. let me think about it a bit.